### DISCUSSION TITLE
Running on Raspberry Pi / Termux

### DISCUSSION THREAD

**Discussion by j-cr** [babashka/babashka#1392] [Q&A]
It would be nice to have prebuilt ARM binaries too, so babashka could be run on a raspberry pi and the likes.

Related links:
https://www.reddit.com/r/Clojure/comments/epowjk/clojure_for_iot_gpio_i2c/
https://github.com/dvlopt/clojure-raspberry-pi

**Comment by borkdude** [comment]
From the GraalVM slack, native-image channel:

<img width="492" alt="Screenshot 2020-01-21 17 26 37" src="https://user-images.githubusercontent.com/284934/72822939-379a8300-3c73-11ea-8df2-36906099bc69.png">

**Comment by j-cr** [comment]
Also, from the reddit discussion:
https://github.com/oracle/graal/issues/1329
so only newer 64-bit ARMs are supported, that means it would work only on raspberry 4 and future versions (but not on raspberry pi zero).

**Comment by borkdude** [comment]
More info:

![Screenshot 2020-02-10 21 07 42](https://user-images.githubusercontent.com/284934/74185667-6344d380-4c49-11ea-8ce4-489a80b7d333.png)

https://github.com/rmh78/quarkus-simple

**Comment by borkdude** [comment]
@j-cr There is now a graalvm-ce-java8-linux-aarch64-20.0.0.tar.gz download available:

https://github.com/graalvm/graalvm-ce-builds/releases

**Comment by borkdude** [comment]
Waiting for someone stepping up to work on this. I'll re-open when there's any new activity.

**Comment by borkdude** [comment]
Which CIs can build for arm64?

- https://sourcehut.org/alpha-details/
- Travis?

**Comment by sogaiu** [comment]
I was able to build and run clj-kondo for aarch64, so I'm now trying babashka.

So far the following line needed to be commented out for the build to proceed:

https://github.com/borkdude/babashka/blob/548ffa851b37c9f0e28e4ae542b9fb72779128c5/script/compile#L20

**Comment by sogaiu** [comment]
![babashka-aarch64](https://user-images.githubusercontent.com/33044872/78732811-99e25600-793b-11ea-88d0-260e00776290.jpg)

**Comment by sogaiu** [comment]
The procedure basically involved:

* appropriate android device
* graalvm java 11 19.3.1 for aarch64
* termux
* https://wiki.termux.com/wiki/Arch
* installing git, gcc, wget, lein within the "prooted" arch distribution (except for lein, use `pacman -S <package>` to install)
* commenting out one line of script/compile
* usual invocation of script/compile

As you can see it takes over 10 minutes on this tablet.

**Comment by sogaiu** [comment]
Some tests fail and I noticed that `(System/getenv "PATH")` also doesn't succeed.

**Comment by sogaiu** [comment]
Here's a zip of the build.

[bb.zip](https://github.com/borkdude/babashka/files/4448856/bb.zip)

**Comment by sogaiu** [comment]
Some additional notes:

* The bb binary would not execute properly outside of the specially prepared environment in which it was built.  There seem to be a variety of restrictions in place on Android.  It ran so-so inside the "proot", but not outside -- the same was true of the java binary that comes with graalvm.

* Currently trying to set up an aarch64 emulated environment (qemu) in which to test the built binary.

* I should be able to provide the emulated environment with more memory too.  The original build was done on a Samsung Tab S4 with 6 GB of memory.

* I thought of trying a build with a Raspberry Pi 2, but a couple of things disuaded me.  First, the default distributions for the RPi are not aarch64-based IIUC.  Also, I think my RPi 2 only has 1 GB of RAM.

* IIUC, the Raspberry Pi 4 has a model that comes with 4 GB and apparently people have had luck making their own aarch64-based distributions, so may be that's another possible path?

**Comment by sogaiu** [comment]
Had some success setting up a qemu environment with aarch64.  The following article was the best that I found on the topic: https://translatedcode.wordpress.com/2017/07/24/installing-debian-on-qemus-64-bit-arm-virt-board/

It was able to execute the bb binary built on the Android device with similar results to how it behaved within the "proot" environment.

It still showed similar errors, but borkdude mentioned that there was a known issue with jdk 11 and he has a fix:

https://github.com/borkdude/clj-reflector-graal-java11-fix

Am currently trying that out.

`project.clj` is being modified by:

* changing the version of `[borkdude/sci.impl.reflector "0.0.1"]` to `[borkdude/sci.impl.reflector "0.0.1-jdk11"]`
* adding the dependency `[borkdude/clj-reflector-graal-java11-fix "0.0.1-graalvm-19.3.1-alpha.2"]` -- this isn't supposed to really go in the value of the top-level :dependencies key, but for the sake of the test, it is being done that way

A complete build on the tablet takes around 20 min.

Building in the qemu environment is quite slow.  Allocating 8 GB of memory doesn't seem to have helped with the speed, but perhaps that is not too surprising.  For subsequent runs I will try increasing the number of cores as well as allocate some more disk space.

**Comment by sogaiu** [comment]
The resulting binary appears to behave better :)

Here it is:

[bb.zip](https://github.com/borkdude/babashka/files/4450513/bb.zip)

**Comment by sogaiu** [comment]
Tangential, but here's clj-kondo:

[clj-kondo.zip](https://github.com/borkdude/babashka/files/4450568/clj-kondo.zip)

**Comment by sogaiu** [comment]
It may be that building would be faster if one tweaks the `-Xmx` setting in the compile script.

**Comment by sogaiu** [comment]
It appears to help a lot (though also added disk space for more swap):
```
[bb:1935]      [total]: 7,017,894.94 ms
```

For comparison, the number for clj-kondo was:
```
[clj-kondo:1049]      [total]: 4,099,406.93 ms
```
These are long times, but the previous attempts did not complete.

So babashka is a bit under 2 hours and clj-kondo is more than 1 hour.

**Comment by sogaiu** [comment]
I tried tweaking `-Xmx` for the tablet but values of 4 and 5 gigs lead to termux being killed by Android if I'm interpreting things correctly.

So it looks like the default of 3g is somewhere close to the limit -- and that means taking in the neighborhood of 20 min.

**Comment by sogaiu** [comment]
@j-cr Regarding use on the RPi models, it does seem to be the case that typical distributions don't use aarch64 for RPi 2 and 3 (notably Raspbian).

It appears though that technically one can run aarch64 based distributions on RPi 2 and 3 (at least scanning some of the forum posts seem to suggest so).  I haven't tried it, but FWIW.

Update: Some RPi 2 are 32-bit and some are 64-bit -- unfortunately, it appears mine is in the former category.

**Comment by sogaiu** [comment]
@borkdude I didn't manage to find a graalvm-ce-java8-linux-aarch64-20.0.0.tar.gz at: https://github.com/graalvm/graalvm-ce-builds/releases (checking via your comment: https://github.com/borkdude/babashka/issues/241#issuecomment-587925422)

Not sure if it was there at one point or not, but FWIW.

**Comment by sogaiu** [comment]
I was able to run an emulated rpi3 via qemu-system-aarch64 -- and the built bb seems to at least start there:
```
Last login: Thu Dec 14 22:10:24 UTC 2017 on ttyAMA0
Linux rpi3 4.14.0-3-arm64 #1 SMP Debian 4.14.12-2 (2018-01-06) aarch64
 
The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.
 
Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
 
Please change the root password by running passwd
root@rpi3:~# mv /boot/firmware/bb /root
root@rpi3:~# cd /root/
root@rpi3:~# ls -al
total 50984
drwx------  2 root root     4096 Dec 14 22:10 .
drwxr-xr-x 21 root root     4096 Jan  8  2018 ..
-rw-------  1 root root      248 Dec 14  2017 .bash_history
-rw-r--r--  1 root root      570 Jan 31  2010 .bashrc
-rw-r--r--  1 root root      148 Aug 17  2015 .profile
-rwxr-xr-x  1 root root 52184576 Apr  8  2020 bb
root@rpi3:~#
root@rpi3:~# uname -a
Linux rpi3 4.14.0-3-arm64 #1 SMP Debian 4.14.12-2 (2018-01-06) aarch64 GNU/Linux
root@rpi3:~# ./bb
Babashka v0.0.81-SNAPSHOT REPL.
Use :repl/quit or :repl/exit to quit the REPL.
Clojure rocks, Bash reaches.
 
user=>
```

**Comment by sogaiu** [comment]
I followed the article here to set up an emulated RPi3 environment in qemu-system-aarch64: https://translatedcode.wordpress.com/2018/04/25/debian-on-qemus-raspberry-pi-3-model/

* the usual pi user didn't seem to be there, so used root
* the typical default password for rpi did work
* used qemu 5.0.0rc2 -- this lead to a gui starting and choosing View -> serial0 helped to see stuff on the screen
* there is no network support for the particular bits I used, so used guestfish to copy bb to one of the partitions (see the article for details -- one uses copy-in instead of copy-out and leave off the --ro flag when starting guestfish)

**Comment by sogaiu** [comment]
Apparently if one installs a 64-bit kernel (and possibly does some configuration tweaking), one may be able to run aarch64 binaries (even without userland being made 64-bit).

https://www.raspberrypi.org/forums/viewtopic.php?f=29&t=250730

If you have such a setup, the binaries that have been posted already would be the ones to try.

**Comment by sogaiu** [comment]
I found this thread that mentions a number of different aarch64-based RPi distributions:

https://raspberrypi.stackexchange.com/questions/100926/64-bit-os-on-raspberry-pi-4/101802#101802

Two that stood out were:

* [Manjaro's 64-bit RPi4 option](https://www.manjaro.org/download/arm/raspberry-pi-4/arm8-raspberry-pi-4-xfce/)
* [Ubuntu's 64-bit RPi3 and RPi4 option](https://ubuntu.com/blog/updated-images-of-ubuntu-for-the-raspberry-pi-2-3-and-4)

If you are more comfortable with Arch-based systems, the Manjaro option should probably be pretty familiar.

If you are more comfortable with Debian-based systems, the Ubuntu option might be one to try.

The thread is probably worth examining as it lists other options (though I urge caution in choice) as well as detailed instructions, tips, etc.

**Comment by iku000888** [comment]
Cheers on the work/research @sogaiu . 
Would be fun if it works on raspberry pis!

**Comment by borkdude** [comment]
Babashka is now built using java11. This means that the tweaks related to java11 in this thread should not be relevant anymore.

**Comment by sogaiu** [comment]
@iku000888 I think someone mentioned at #babashka on slack that they were going to try with a Raspberry Pi 4.  Not sure how that turned out.

**Comment by borkdude** [comment]
Interesting article about compiling Java to native with GraalVM for Rasbperry:

https://quarkus.io/blog/quarkus-native-on-a-raspberry-pi/

**Comment by harryvederci** [comment]
@sogaiu the binaries you shared here work on my Raspberry Pi 4 (aarch64, kernel release 5.4.75-v8+).

I'd like to be able to build them myself from that device as well, though.
I'm not familiar with the process, and reading the above I don't think I'd be able to reproduce what you did (example: "commenting out one line of script/compile").

Did you do it manually, or do you have a script that I could use for this purpose?
If not: I'm really happy with the binaries, so thanks :)

**Comment by sogaiu** [comment]
@harryvederci Glad to hear things are working for you :)

Sorry I don't have a script -- however, I did note which line I commented out:

https://github.com/borkdude/babashka/issues/241#issuecomment-610688377

It might have been better to have put all of the pieces in one place -- my apologies wrt that.

I didn't check out the article borkdude mentioned: https://github.com/borkdude/babashka/issues/241#issuecomment-721634861

It could be that things are a lot easier now.  May be it's worth looking at that?

**Comment by brdloush** [comment]
After reading this discussion, I wasn't really sure if it's possible to get a recent pre-compiled babashka for my raspberry 4b (aarch64). Or at least how easy/hard is it to build the native image myself. 

TLDR: it's surprisingly easy to build `bb` on real raspberry pi 4 :+1: :clap: 

Since I didn't find any recent pre-built aarch64 binaries in downloads, I tried to compile one myself. 

I was building babashka directly on my raspberry pi 4, running ubuntu 20.04.1 LTS 64bit:

```bash
ubuntu@ubuntu:~/projects/babashka$ cat /proc/cpuinfo  | grep Model
Model		: Raspberry Pi 4 Model B Rev 1.4
```

```bash
ubuntu@ubuntu:~/projects/babashka$ uname -a
Linux ubuntu 5.4.0-1026-raspi #29-Ubuntu SMP PREEMPT Mon Dec 14 17:01:16 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux
```

```bash
ubuntu@ubuntu:~/projects/babashka$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

Compilation was easy. I only had one issue: missing `libz.a` file. After installing the missing dependency (I believe `libz-dev` did the job), the compilation worked like a charm! Here's I believe all the simple steps I had to do:

```bash
# get gcc, jdk, libz library
sudo apt install openjdk-11-jdk libz libz-dev build-essentials rlwrap curl

# get fresh copy of graalvm including native-image
mkdir ~/app ~/projects
cd ~/app
wget https://github.com/graalvm/graalvm-ce-builds/releases/download/vm-21.0.0/graalvm-ce-java11-linux-aarch64-21.0.0.tar.gz
tar xf graalvm-ce-java11-linux-aarch64-21.0.0.tar.gz
rm graalvm-ce-java11-linux-aarch64-21.0.0.tar.gz
export GRAALVM_HOME=/home/ubuntu/app/graalvm-ce-java11-21.0.0 # better to add this to ~/.bashrc
export PATH=$PATH:$GRAALVM_HOME # better to add this to ~/.bashrc
gu install native-image

# clone babashka and its child modules
cd ~/projects
git clone https://github.com/borkdude/babashka --recursive
cd babashka

# build uberjar
time ./script/uberjar
# real	1m53.673s
# user	4m29.222s
# sys	0m24.092s

# build native bb image
time ./script/compile # this takes a quite a long time (~ 20 minutes), but it's totally worth it...
# real	19m10.409s
# user	70m35.943s
# sys	0m29.795s
```

After compilation, `bb` binary seems to work just fine. Even my non-trivial babashka scripts which I use on my regular x86 amd64 laptop are working correctly. And `bb` itself is pretty fast on PI4! Well done! :clap: 

```bash
ubuntu@ubuntu:~/projects/babashka$ ./bb --version
babashka v0.2.8-SNAPSHOT
```

```bash
ubuntu@ubuntu:~/projects/babashka$ time ./bb -e ':hello'
:hello

real	0m0.044s
user	0m0.013s
sys	0m0.033s
```

```bash
ubuntu@ubuntu:~/projects/babashka$ time ./bb -e '(-> (curl/get "https://postman-echo.com/get?foo=1") :body (json/parse-string true) :headers clojure.pprint/pprint)'
{:x-forwarded-proto "https",
 :x-forwarded-port "443",
 :host "postman-echo.com",
 :x-amzn-trace-id "Root=1-6008a83d-361da9650e87109f40d92ff1",
 :user-agent "curl/7.68.0",
 :accept "*/*",
 :accept-encoding "deflate, gzip, br"}

real	0m0.827s
user	0m0.099s
sys	0m0.070s
```

**Comment by borkdude** [comment]
@brdloush This is awesome! Do you think it would be worthwhile to distribute the Raspberry binary? I'm not sure how portable these binaries are - once compiled with GraalVM linux aarch64, they can run on any other Raspberry?

**Comment by brdloush** [comment]
As for the compatibility... well, I grabbed the binary that @sogaiu made and published in one of the previous posts (https://github.com/babashka/babashka/issues/241#issuecomment-610938199) and it worked without issues. I'd expect that if you grab that binary built for aarch64, it should be possible to run it on any device running 64bit arm cpu running 64bit linux OS. @sogaiu was using it on android tablet, so I believe the binary may be considered "quite portable" :smile: 

For me babashka on raspberry makes perfect sense. Both It's startup time and runtime performance looks really good so far. I wonder... is there any bb benchmark that I could run on it?

I'm really looking forward to hacking some bb scripts that will just scrape some stuff from the web, parse it, process it and store it to some local storage. It will benefit from the fact that such script will be running on a tiny cheap, but relatively powerful computer which consumes really small amount of electricity and has no fans. In other words: it can run 24/7 in my bedroom. 

If I'm not mistaken (?), currently just the 64-bit ARM native-images can be built by `native-image`. Which might be a problem for older Raspberries.. Most would probably be running Raspberry Pi OS (32-bit), Raspbian (32-bit) or Ubuntu (32-bit). The reason behind this is that older models of raspberries used to have smaller amount of memory (RPi3 still had only 1GB!). But with RPi4, there are 1GB, 2GB, 4GB and even 8GB options. And 8GB one costs around 75-100 USD. Running 64-bit OS on 8GB is a reasonable option I think.

According to list of distributions offered by the official `Raspberry Pi Imager v1.5` tool, it seems that one can use 64-bit Ubuntu on RPi3, RPi4 and Rpi400.

**Comment by borkdude** [comment]
Feel free to zip the binary and upload it here for other people to try, as an "unofficial" distribution, like what @sogaiu did. 

A tip for faster compilation: you can disable some of the libraries, if you aren't using most of them:

https://github.com/babashka/babashka/blob/master/doc/build.md#feature-flags

The `BABASHKA_LEAN` flag only keeps the most essential libs.

**Comment by brdloush** [comment]
Ok, here's the zipped aarch64 binary:

[bb-aarch64-2021-01-20_git_rev_14c33625ecb203af2513eb5f527db63e77848ae5.zip](https://github.com/babashka/babashka/files/5845766/bb-aarch64-2021-01-20_git_rev_14c33625ecb203af2513eb5f527db63e77848ae5.zip)

As for LEAN babashka: thanks for the tip. But 20 minutes in not that bad for building the real thing. But I'll try it, just for fun :)

**Comment by brdloush** [comment]
just FYI:

LEAN native image gets built in `16m12s` (compared to `19m10.409s` for default `bb`). It results in 56MB `bb` binary, default `bb` is 70M. From the startup performance perspective: they seem to start up in the same time.

**Comment by harryvederci** [comment]
> Ok, here's the zipped aarch64 binary:

Works on my machine!

**Comment by den1k** [comment]
Incredibly exciting to imagine the possibilities of running babashka + datalevin on raspberry pi

And impeccable timing as @huahaiy is working on the missing piece: https://github.com/juji-io/datalevin/issues/6

**Comment by borkdude** [comment]
This is great. We can make a datalevin pod, like we already have SQL pods:

- https://github.com/babashka/babashka-sql-pods/
- https://github.com/babashka/pod-babashka-sqlite3

and/or add it as a feature flag and then that should work when also compiled on linux/aarch64.

**Comment by eigenhombre** [comment]
Super excited to see ARM progress for RPi.  Question: I have a new Pi running Raspbian, which I think is still 32-bit.  Is the new ARM Babashka 64-bit only?

**Comment by borkdude** [comment]
@eigenhombre Babashka is compiled using GraalVM and there is no 32-bit release of native-image for aarch.

**Comment by brdloush** [comment]
@eigenhombre you might want to look at [this article](https://medium.com/for-linux-users/how-to-make-your-raspberry-pi-4-faster-with-a-64-bit-kernel-77028c47d653). It seems that it's possible to use 32-bit raspbian (so all the installed libraries are 32-bit) and switch to 64-bit kernel. But be sure to backup before trying.

Also first make sure that your raspberry's CPU is actually 64-bit. If it's PI3 or PI4, it should be. Not sure about previous PIs.

**Comment by borkdude** [comment]
This babashka binary should work on RaspberryPi 64bit. It's built on CircleCI now!

https://15839-201467090-gh.circle-artifacts.com/0/release/babashka-0.2.14-SNAPSHOT-linux-aarch64.tar.gz

Anyone wants to try?

**Comment by brdloush** [comment]
Works on my RPI4 

```bash
$ time ./bb -e ':hello'
:hello

real	0m0.045s
user	0m0.008s
sys	0m0.039s
```

**Comment by harryvederci** [comment]
Works on my Chromebook as well!

Bedankt/thanks Michiel!

**Comment by borkdude** [comment]
What does `uname -m` print for you? `aarch64`? I can use this to update the install script.

**Comment by harryvederci** [comment]
Exactly that on the Chromebook: `aarch64`

**Comment by borkdude** [comment]
Thanks! I updated the `install` script so it should recognize your architecture and install babashka correctly on aarch64.

**Comment by harryvederci** [comment]
Works! Thanks a lot!

**Comment by hamann** [comment]
I did the kernel upgrade from 32-bit to 64-bit today as mentioned in https://github.com/babashka/babashka/issues/241#issuecomment-764899056, but for me bb isn't working 

```sh
~ pi@raspberrypi
❯ grep Model /proc/cpuinfo
Model		: Raspberry Pi 4 Model B Rev 1.2

~ pi@raspberrypi
❯ uname -a
Linux raspberrypi 5.10.17-v8+ #1403 SMP PREEMPT Mon Feb 22 11:37:54 GMT 2021 aarch64 GNU/Linux

~ pi@raspberrypi
❯ file /usr/local/bin/bb
/usr/local/bin/bb: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=0dbe834f277b844079eba279ecd50988d91e213e, for GNU/Linux 3.7.0, with debug_info, not stripped

~ pi@raspberrypi
❯ /usr/local/bin/bb
zsh: no such file or directory: /usr/local/bin/bb
```

I guess a 64-bit userland is also required.

**Comment by borkdude** [comment]
@hamann Can you run `ldd /usr/local/bin/bb` and show what is says? Maybe you are missing libstdc++

**Comment by hamann** [comment]
`ldd`, of course 🤦‍♂️

```
❯ ldd /usr/local/bin/bb
	not a dynamic executable

❯ strace /usr/local/bin/bb
execve("/usr/local/bin/bb", ["/usr/local/bin/bb"], 0xffc4c6d0 /* 52 vars */) = -1 ENOENT (No such file or directory)
strace: exec: No such file or directory
+++ exited with 1 +++
```

**Comment by borkdude** [comment]
@hamann You have downloaded the static aarch64 executable? Can you try the normal aarch64 one?

**Comment by hamann** [comment]
@borkdude I tried bb with the installer and also downloaded `babashka-0.3.0-linux-aarch64.tar.gz` manually from published releases. I there another one I don't see?

**Comment by borkdude** [comment]
Oh correct, there is no static aarch64 build yet. It might be an issue with your OS install then, like you said.

**Comment by borkdude** [comment]
@hamann There was a bug in the installer script. Could you maybe try again? However, manually downloading the tar.gz should have worked.

**Comment by hamann** [comment]
@borkdude nope, same result.

I decided to reinstall the pi with a 64-bit OS and finally bb works 👌

**Comment by borkdude** [comment]
Notes from @brdloush about running bb on Android tablet and phone that may come in handy:
 
Tomas Brejla  FYI that M2-801W tablet is running rather ancient Android 6. Simple installation of termux app from google play and copying statically compiled bbwas enough to get it running.
For newer androids, (I have Xiaomi MI8 phone with Android 11 10) , it might be a bit more problematic due to newly introduced changes in security.
Yet, in the end, I was able to run bb even on that phone. All I had to additionally do was to to install proot package (apt install proot) and then execute termux-chroot script before actually trying to run bb . Without these adjustments, bb wouldn't start on this device.
It's wonderful what can be achieved (and in relatively easy way) these days.
I wonder what use-case scenarios people will come with. Having the ability to run full-blown bb in that powerful device resting in your pocket, may spark many new ideas.
Tomas Brejla  Althought it should be noted that Termux's position is rather complicated these days, as explained in this article
https://www.xda-developers.com/termux-terminal-linux-google-play-updates-stopped/

**Comment by naxels** [comment]
> @borkdude nope, same result.
> 
> I decided to reinstall the pi with a 64-bit OS and finally bb works 👌

Compiling the uberjar on the 32-bit Raspberry Pi OS will ultimately work after installing Leiningen & OpenJDK, however I did the same, just installed the beta 64-bit OS and the installer just works as expected with the aarch64, will have to manually sudo move the bb file though.
bb is fast now on Raspberry Pi 3 :)

**Comment by borkdude** [comment]
@naxels Thanks for letting me know!

**Comment by akielbowicz** [comment]
hi, with the help from Ashish in Slack, I manage to run babashka on Termux (Android 15). 

Here are the steps to install it:

0. As a prerequisite you need to install [glibc-runner](https://github.com/termux/glibc-packages)
	```sh
	pkg install glibc-repo -y
	pkg install glibc-runner -y
	```

1. Download the [installer script](https://github.com/babashka/babashka#installer-script)
	```sh
	cd $HOME
	curl -sLO https://raw.githubusercontent.com/babashka/babashka/master/install
	chmod +x install
	```

3. Install babashka into a known directory, I choosed `$HOME/babashka`
	```sh
	mdir $HOME/babashka
	./install --dir "$HOME/babashka"`
 	```

4. Now you can run babashka with
	```sh
	grun -f "$HOME/babashka/bb"
	```

	```sh
	Message from glibc-runner: searching for libraries...
	Message from glibc-runner: searching libraries was successful
	Babashka v1.12.196 REPL.
	Use :repl/quit or :repl/exit to quit the REPL.
	Clojure rocks, Bash reaches.

	user=>
	```

Hope that this helps

<details>
<summary>More details on the issue</summary>

I was getting the following message when trying to run it:

```sh
~ $ ./babashka/bb
bash: /data/data/com.termux/files/home/babashka/bb: cannot execute: required file not found
```

Ashish figured that it required some `glibc` packages to run

```sh
~ $ ldd "$HOME/babashka/bb"
        libdl.so.2 => not found
        libpthread.so.0 => not found
        libc.so.6 => not found
        ld-linux-aarch64.so.1 => not found
```

And installing them fixed the issue.

#### Some extra bits of info to understand the issue:

There are multiple implementations of C Standard Library (`libc`), ie, components that talk to the kernel. One of the most common implementations is [GNU glibc](https://en.wikipedia.org/wiki/Glibc), and Android uses its own implementation [Bionic libc](https://en.wikipedia.org/wiki/Bionic_(software)) which has some differences in the components. 

#### Configuring bb to run with `grun`

I'm not sure how stable this is or if it will break something so I didn't recommend it as the main way to install it.

3. To avoid configuring extra stuff, install bb directly on the users bin directory, you can check it on $PATH, or using the $PREFIX (mine is, '/data/data/com.termux/files/usr'
	```sh
	./install "$PREFIX/bin"
	```
4. Configure the binary to run
	```sh
	grun --configure "$PREFIX/bin/bb"
	```

5. Check that now it finds the linked libraries
	```sh
	$ ldd "$PREFIX/bin/bb"
    libdl.so.2 => /data/data/com.termux/files/usr/glibc/lib/libdl.so.2
    libpthread.so.0 => /data/data/com.termux/files/usr/glibc/lib/libpthread.so.0
    libc.so.6 => /data/data/com.termux/files/usr/glibc/lib/libc.so.6
    ld-linux-aarch64.so.1 => /data/data/com.termux/files/usr/glibc/lib/ld-linux-aarch64.so.1
	```

6. If you run `bb` you might get an error

	```sh
	~ $ bb
	bb: error while loading shared libraries: libdl.so: cannot open shared object file: No such file or directory
 	``` 

	According to the [glibc wiki](https://github.com/termux-pacman/glibc-packages/wiki/About-glibc-runner-(grun)#configures-the-shell-to-make-glibc-binaries-work) you can fix it by unsetting a variable. 

	```sh
	unset LD_PRELOAD
	```
	this is not persisted so you need to unset it every time you start a new Termux session. Maybe it's safe to add it to an rc or .profile file, but I'm not feeling adventurous yet. Setting and alias might work too `alias bbk='LD_PRELOAD="" bb'` 
	

7. Now you can run `bb` directly as it's on your path
	
	```shell
	$ bb
	Babashka v1.12.196 REPL.
	Use :repl/quit or :repl/exit to quit the REPL.
	Clojure rocks, Bash reaches.

	user=>
	``` 

</details>

**Reply by rene-descartes2021** [reply]
Thanks for sharing it works with glibc-runner on Termux, not involving a proot. Hopefully no consequent problems.

In looking at GraalVM for Android support, they advised using [Gluon Substrate](https://github.com/gluonhq/substrate) to provide compiled natives for Android and other platforms.

**Reply by sogaiu** [reply]
It worked for me too :+1: 

Some of the details in the "More details on the issue" seemed to be significant here.  Without them (at the very least, configuring with `grun` and `unset LD_PRELOAD`), I'm not sure I would have gotten it working :sweat_smile: 

Also, FWIW, I got a different error in step 6 (something about an invalid ELF header), though the suggested command for that step did fix things here.

---

There does seem to be a downside to the `grun` approach though.  Once `unset LD_PRELOAD` has been invoked, I don't seem to be able to start `clj` or `clojure` (which happens to be something that [the `bb` script I want to run does](https://github.com/clj-commons/rewrite-clj/blob/9f871e01281f278be55a945e62e6fc2054e324be/script/dev_repl.clj#L58))...
